### DISCUSSION TITLE
How do I enable turn server to continue serving users that have been connected before the restart?

### DISCUSSION THREAD

**Discussion by GatlinHa** [coturn/coturn#1159] [Q&A] [answered]
My turn server is working and it has been connected by some users that they are on a voice or video call. Now I need to upgrade turn server to a new version due to some problems, and it may only take a few seconds to restart. But when I restart, the previously connected user will be disconnected.This is a bad use experience. Is there a way around to solve this problem?

I have some ideas. I want to save all the information of Allcation, Permission and ChannelBind in DB and synchronize the memory with DB in real time. When the server restarts, the data in the DB is reloaded to try to restore the previous connection data. Are there better ways, or has been implemented ways?

**Comment by eakraly** [comment, marked as answer]
Hi @GatlinHa 

Great idea but it has a few drawbacks.
- Adding DB for this has huge development and operational overhead (think code, bugs, DB, permissions, monitoring etc)
- Someone needs to write that code
- You also need to store SSL context because some clients may be using TLS or even DTLS (and this is not trivial)

Zooming out of this question - statefulness of turn sessions (as well as VoIP/WebRTC and anything that uses TURN in general) is a common problem. It is not only about upgrade but also about scaling in (when you want to shut down unused servers)

 People solve it by "draining" servers before shutting down/restart. The process is following:
- You select a server you decide to shutdown (or restart for upgrade)
- You stop sending new users to it (stop allocating new sessions) but keep serving existing sessions - this is draining (no new sessions but existing sessions finish gradually)
- Continue monitoring the metrics until the load is 0 (or low enough) 
- Restart/upgrade/shutdown
- If that was upgrade/restart - now return the server to the allocation pool so it can start handling traffic.
- Repeat with next server

To accomplish this you would need:
- Monitoring (prometheus or redis)
- Some control over allocation/load balancing of turnserver (could be LB, or DNS, or your application logic)
